### PR TITLE
Release workflow

### PR DESCRIPTION
**Pull request by acalcutt**
*Please note, this is untested right now and I am giving it for review and/or ideas for https://github.com/mapbox/node-pre-gyp/issues/739.*

This is a release workflow for node-pre-gyp. It is based on the node release workflow for [maplibre-native](https://github.com/maplibre/maplibre-native) at https://github.com/maplibre/maplibre-native/blob/main/.github/workflows/node-release.yml which I worked on getting working.

The idea behind this was, someone needs to approve and add changelog for each release, which should be a PR. When the version in package.json changes, this release workflow will check if that version has been published. If it has not published it, it will do that and also create a github release. the github release will also contain the text of the CHANGELOG.md for the version being published.

The workflow is also made to deal with pre-releases, so if the version is a pre-release, it will publish it as the next npm release

Whoever creates the PR for a new release would follow the instructions in
[RELEASE.md](https://github.com/mapbox/node-pre-gyp/pull/744/files#diff-2b1b69303b927a484e02c7fad9fc87d0d3ff0dc22ae1da0ecd0dc935d922a23c)

Needs 'NPM_TOKEN' Repository secret. A secret that contains a user npm token with permission to publish

### CODE REVIEW

**Comment by cclauss** [comment]
@mapsam Given all the progress that has been made in the past 24 hours, I believe that we need a ___provisional___ release of `node-pre-gyp` to https://www.npmjs.com/package/@mapbox/node-pre-gyp. This _pre-release_ would allow multiple parties to kick the tires (for a week or two?) ___before___ we make a general release.

You or one of your colleagues would need to work with @acalcutt to add the ___required___ `NPM_TOKEN` as a repository secret.

**Comment by acalcutt** [comment]
Just an FYI, if you did follow my instructions, when running the first command, and 'pre' version will make a pre-release/next release when published.  So with a version incremted like
`npm version preminor --preid pre --no-git-tag-version` 
would bring this package to 1.1.0-pre.0 from the current 1.0.11 and the workflow would publish it as a pre-release.

**Comment by acalcutt** [comment]
I gave this a test on my fork, and fixed a few issues and make it more similar to the ci. it should now be working correctly. I used it to make this test release

https://www.npmjs.com/package/@acalcutt/node-pre-gyp/v/2.0.0-pre.1
https://github.com/acalcutt/node-pre-gyp/releases/tag/v2.0.0-pre.1

it was made in this workflow run
https://github.com/acalcutt/node-pre-gyp/actions/runs/9735220985

**Comment by mapsam** [comment]
Nice work y'all 🙌  happy to get `NPM_TOKEN` added as a repo secret for the `@mapbox/node-pre-gyp` module. I'll report back when that has been added.

**Comment by mapsam** [comment]
@acalcutt @cclauss unsurprisingly, it's a little hard for me to get a static NPM token for our `@mapbox/` namespace packages. Internally we're working on a better solution for public repos + npm + outside maintainers that will resolve this. For now I've set a reminder to refresh the token periodically before it is set to expire.

A couple questions:

* do you need me to modify any of the branch protections for the `master` branch? Currently no one is allowed to manually push to that branch
* are there protections to enforce only repo maintainers can trigger a workflow? Just want to be extra cautious outside contributors and forks cannot run the workflow themselves

**Comment by mapsam** [comment]
I've added an `NPM_TOKEN` secret to the repo 👍

**Comment by acalcutt** [comment]
`do you need me to modify any of the branch protections for the master branch? Currently no one is allowed to manually push to that branch`
This does not require any changes to branch protections. 

`are there protections to enforce only repo maintainers can trigger a workflow? Just want to be extra cautious outside contributors and forks cannot run the workflow themselves`

The only way to trigger this would be to push a version change to the master branch on this repo into package.json. So either by a PR, which a maintainer would have to approve, or by a commit by someone who has admin rights to bypass branch protection.

If someone clones this repo, the workflow will not be able to publish because a valid 'NPM_TOKEN' will not exist in the new repository (it would only exist in this one). They could add their own NPM_TOKEN into their repo, but they would have to have permissions to update the package or it fails.

**Comment by mapsam** [comment]
Thanks for the quick response @acalcutt!

**Comment by acalcutt** [comment]
For example, to publish my test version, I had to do 3 things

1.) change my package.json package name
https://github.com/acalcutt/node-pre-gyp/blob/master/package.json#L2

2.) change the workflow to check my version of the package
https://github.com/acalcutt/node-pre-gyp/blob/master/.github/workflows/release.yml#L26

3.) Provide my own NPM_TOKEN  with my own npm key, that hass permission to @acalcutt/node-pre-gyp

**Comment by cclauss** [comment]
I agree that the branch protections should not be changed.

@acalcutt Would you be interested in becoming a maintainer of this repo?

**Comment by acalcutt** [comment]
I don't mind submitting PRs once and a while or helping out where I can, but I'm not sure how much of a maintainer I want to be. The things I have submitted so far are mainly stuff I know from trying to keep this limping along for my own projects.

I've mainly submitted these changes in hopes of seeing some up to date releases so I no longer need to maintain @acalcutt/node-pre-gyp. However in my current maplibre-native release I still publish node 16, so to move to this I would have to officially drop it (which i plan to do soon anyway), so using this package would be a breaking change once it is published.

**Comment by cclauss** [comment]
@mapsam Should this be merged or should something else be done in its place?

**Comment by mapsam** [comment]
@cclauss if the workflow works for you, it works for me! 

My initial questions were just focused on blast-radius security because I wasn't able to get a well-scoped NPM token. I've since been able to generate an `NPM_TOKEN` only scoped to _just_ the `@mapbox/node-pre-gyp` module, so my concerns there are settled. That token has been updated in the repo and should be available to test with now 👍

**Comment by acalcutt** [comment]
My feeling is it has been tested and works and can always be modified more
in the future if another approach is found. It seems worth giving it a try.

File: RELEASE.md
Comment (benmccann) [inline RELEASE.md:3]: too many commas :laughing: 
```suggestion
1. Change the version number in `package.json`. Run the following command in the package root directory, replacing <update_type> with one of the semantic versioning release types (prerelease, prepatch, preminor, premajor, patch, minor, major):
```

File: RELEASE.md
Comment (benmccann) [inline RELEASE.md:5]: you could put this in triple backticks (there's no way for me to suggest that on github since it also uses triple backticks for suggestions :laughing:)

**Comment by benmccann** [comment]
@cclauss should we merge this and test it out to release a `2.0.0-rc.1` or something along those lines?

**Comment by acalcutt** [comment]
Sorry for the late reply. Let me know if you want me to make any more wording changes. 

I am still around to help if this gets merged and there are any issues.

File: RELEASE.md
Comment (acalcutt) [reply RELEASE.md:6]: I added in the ticks

**Comment by cclauss** [comment]
Please move forward with this @benmccann

**Comment by acalcutt** [comment]
by the way, after merging this, if you wanted to do a 2.0 prerlease from the current 1.x release, you would do the following in a PR (or directly in the repo if you are an admin)

1.) Bump the version to 2.0.0-pre.0
`npm version premajor --preid pre --no-git-tag-version`

2.) Add a changelog with the '2.0.0-pre.0' heading

3.) Commit the changes, so the workflow can pick it up

After that, you would wanted a '2.0.0-pre.1', you would do the same thing with 
`npm version prerelease --preid pre --no-git-tag-version`

Or if you wanted a non-pre release of 2.0.0
`npm version major--preid pre --no-git-tag-version`


I know '2.0.0-rc.1' was mentioned above, but this isn't set up to make a pre-release for 'rc' only 'pre'

File: RELEASE.md
Comment (benmccann) [inline RELEASE.md:8]: let's add some of the details from your last comment to the docs, so that they don't get lost and we don't forget how to cut a release :smile: 
```suggestion

`--preid` specifies which suffix to use in the release such as `pre`, `next`, `beta`, `rc`, etc.

`prepatch`, `preminor`, and `premajor` start a new series of pre-releases while bumping the patch, minor, or major version. E.g. `premajor` with `--preid pre` would do a prerelease for a new major using the `-pre` suffix (i.e. it would be a new major with `-pre.0`)
```

**Comment by benmccann** [comment]
> After that, you would wanted a '2.0.0-pre.1', you would do the same thing with
npm version prerelease --preid pre --no-git-tag-version

How does it know what version number to use if you've done a `2.0.0-pre.0` and `1.12.0-pre.0`? Is it based off the `package.json` for the current branch?

> Or if you wanted a non-pre release of 2.0.0
npm version major--preid pre --no-git-tag-version

Do you need to specify `--preid pre` here?

**Comment by acalcutt** [comment]
`How does it know what version number to use if you've done a 2.0.0-pre.0 and 1.12.0-pre.0? Is it based off the package.json for the current branch?`

The workflow reads it from the package.json from the branch it is running in. So if you submit a PR to madter, this workflow in master would pick it up. The workflow is set to run on [push to the master branch ](https://github.com/mapbox/node-pre-gyp/pull/744/files#diff-87db21a973eed4fef5f32b267aa60fcee5cbdf03c67fafdc2a9b553bb0b15f34R5).

To run it on another branch, I usually just change [this line](https://github.com/mapbox/node-pre-gyp/pull/744/files#diff-87db21a973eed4fef5f32b267aa60fcee5cbdf03c67fafdc2a9b553bb0b15f34R5) to the new branch in the workflow file on that branch You can have two branches with different version number and it will work, provided the version being published doesn't already exist on npm (it it does the workflow will not run because the publish check will fail at https://github.com/mapbox/node-pre-gyp/pull/744/files#diff-87db21a973eed4fef5f32b267aa60fcee5cbdf03c67fafdc2a9b553bb0b15f34R22-R31

`Do you need to specify --preid pre here?`
I don't think it has to be be pre here, as long as it gets detected as a pre-release by this code in the workflow https://github.com/mapbox/node-pre-gyp/pull/744/files#diff-87db21a973eed4fef5f32b267aa60fcee5cbdf03c67fafdc2a9b553bb0b15f34R66-R74
(in the maplibre node project were I created this code for it matters because I am also using node-pre-gyp-github, which I added [code to detect the 'pre'.](https://github.com/acalcutt/node-pre-gyp-github/blob/master/index.js#L64)... but for this project I guess it really doesn't matter as long as the workflow command picks it up as a pre-release)

File: RELEASE.md
Comment (benmccann) [inline RELEASE.md:12]: adding those just discussed details
```suggestion

You can use `prerelease` to bump the version for a new pre-release version. E.g. you could run `npm version prerelease --preid pre --no-git-tag-version` to go from `-pre.0` to `-pre.1`.

For regular versions, you can use `patch`, `minor`, or `major`. E.g. `npm version major --no-git-tag-version`.

2. Update the changelog, which can be found in `CHANGELOG.md`. The heading must match `## <VERSION>` exactly, or it will not be picked up. For example, for version 1.0.11:
```

**Comment by acalcutt** [comment]
I rebased this and ran a test in the forked repo. To test this I did the following

1.) In my forked repo I added my own npm key to the 'NPM_TOKEN' in repository secrets.
image

2.) Because my npm key does not have write access to @mapbox/node-pre-gyp, I changed this so it would publish to @acalcutt/node-pre-gyp-test . I also changed the workflow to run in my branch 'test_release_12-7-2024'. Thoase changes look like this
https://github.com/WifiDB/node-pre-gyp/commit/e8e0ddec160a6a10e78aa7e5cc2e8ae75694ba6c

3.) I made a commit that changed the version number and updated the changelog header. I tested using 'rc' , so the command looked like
`npm version premajor --preid rc --no-git-tag-version`
https://github.com/WifiDB/node-pre-gyp/commit/b839bb5ce426d6345d35c743f4289c5bd8447492

4.) On commit, the workflow ran and published the changes
https://github.com/WifiDB/node-pre-gyp/actions/runs/12214664120

The github release got published with release notes
https://github.com/WifiDB/node-pre-gyp/releases/tag/v2.0.0-rc.1

On npm, the release was published as a 'next' release
https://www.npmjs.com/package/@acalcutt/node-pre-gyp-test/v/2.0.0-rc.1?activeTab=versions

**Review by benmccann — APPROVED** [review]
ok, let's try it!

**Comment by acalcutt** [comment]
Seems like it needs a plugin allowed (or changed to a supported version)

https://github.com/mapbox/node-pre-gyp/actions/runs/12214984091
martinbeentjes/npm-get-version-action@v1.3.1 is not allowed to be used in mapbox/node-pre-gyp. Actions in this workflow must be: within a repository that belongs to your Enterprise account, created by GitHub, or matching the following: anchore/*, aws-actions/*, bridgecrewio/*, cloudposse/*, docker/*, mapbox/*, octokit/*, smartlyio/*, 8398a7/action-slack@v3, amannn/action-semantic-pull-request@v5*, aquasecurity/trivy-action@0.12.0, arduino/setup-protoc@v1, ArtiomTr/jest-coverage-report-action@v2*, benchmark-action/github-action-benchmark@v1.14.0, codecov/codecov-action@v*, cycjimmy/semantic-release-action@v4*, dbt-checkpoint/dbt-checkpoint@v1*, gitleaks/gitleaks-action@v2, grafana/k6-action@v0.2.0, jenseng/dynamic-uses@v1, jwalton/gh-find-current-pr@v1*, ludeeus/action-shellcheck@2.0.0, ncipollo/release-action@v1, peter-evans/create-pull-request@v*, prisma-cloud-shiftleft/iac-scan-action@v*, ruby/setup-ruby@v*, slackapi/slack-github-action@v*, thollander/actions-comment-pull-request@...

**Comment by benmccann** [comment]
It looks like that action does almost nothing. Could we just copy that code into our workflow and remove use of that action?
https://github.com/martinbeentjes/npm-get-version-action/blob/main/entrypoint.sh

If you're up for sending a PR, I can merge it.

**Comment by acalcutt** [comment]
The action pulls the version number out of the commit, which is used in the changelog and github publish steps steps (as steps.package-version.outputs.current-version in https://github.com/mapbox/node-pre-gyp/blob/3eba8e4c1f602f82e79ed0e15f01f1307cfb74b8/.github/workflows/release.yml#L66-L91 )

**Comment by acalcutt** [comment]
It could possibly be modified to use the one pulled above in https://github.com/mapbox/node-pre-gyp/blob/3eba8e4c1f602f82e79ed0e15f01f1307cfb74b8/.github/workflows/release.yml#L28 like the prerelease tag does

**Comment by benmccann** [comment]
computing `current-version` is a 1-liner:
```
CURRENT_VERSION=$(cat ${PACKAGE_JSON_PATH}/package.json | jq '.version' | tr -d '"')
```

**Comment by acalcutt** [comment]
I added https://github.com/mapbox/node-pre-gyp/pull/887 to remove that action

**Comment by benmccann** [comment]
Awesome! Looks like it's working now as far as I can tell. Mind sending a PR for the release? I can't merge my own PR, but could merge yours :smile: 

I think we need to run `npm version premajor --preid rc --no-git-tag-version` and update `Unreleased - 2.0.0` to `2.0.0-rc.0` in the `CHANGELOG.md`. I also noticed a typo for "automatically" on the last line of `RELEASE.md` that we could try to update at the same time

**Pull request merged**
